### PR TITLE
perf(locator): prepare/commit for groups refresh

### PR DESCRIPTION
**Pull request by 3Hren**
This commit decomposes the prepare and broadcast logic for
massive routing groups refresh.

When we iterate over outgoing streams to notify each there
is no need to rebuild routing map every iteration.

Also the bool `replace` flag is replaced with named enum.

Presence of boolean flags in method signatures is anti-pattern, because
when you invoke such method you should always switch to its
declaration to match what that `true` or `false` value actually
means.

### CODE REVIEW

**Comment by 3Hren** [comment]
@kobolog @noxiouz @antmat PTAL

**Comment by 3Hren** [comment]
This PR is partially conflicts with #242, they cannot be both merged without rebase.

File: src/service/locator.cpp [hunk 1]
```diff
@@ -456,7 +487,7 @@ locator_t::uuid() const {
 
 results::resolve
 locator_t::on_resolve(const std::string& name, const std::string& seed) const {
-    const auto remapped = m_rgs.apply([&](const rg_map_t& mapping) -> std::string {
+    const auto remapped = routing.groups.apply([&](const rg_map_t& mapping) -> std::string {
```
Comment (antmat) [inline src/service/locator.cpp]: How about moving this logic inside routing_t?

File: src/service/locator.cpp [hunk 2]
```diff
@@ -524,7 +555,7 @@ locator_t::on_refresh(const std::vector<std::string>& groups) {
     const auto storage = api::storage(m_context, "core");
     const auto updated = storage->find("groups", std::vector<std::string>({"group", "active"}));
 
-    m_rgs.unsafe() = boost::accumulate(groups, *m_rgs.synchronize(),
+    routing.groups.unsafe() = boost::accumulate(groups, *routing.groups.synchronize(),
```
Comment (antmat) [inline src/service/locator.cpp]: and this

File: src/service/locator.cpp [hunk 3]
```diff
@@ -550,16 +581,18 @@ locator_t::on_refresh(const std::vector<std::string>& groups) {
...
-    const auto ruids = boost::accumulate(*m_routers.synchronize(), ruid_vector_t{},
+    const auto ruids = boost::accumulate(*routing.routers.synchronize(), ruid_vector_t{},
         [](ruid_vector_t result, const router_map_t::value_type& value) -> ruid_vector_t
     {
         result.push_back(value.first); return result;
     });
 
+    const auto results = routing.dump();
+
     for(auto it = ruids.begin(); it != ruids.end(); ++it) try {
```
Comment (antmat) [inline src/service/locator.cpp]: I would suggest brackets here

Comment (3Hren) [inline src/service/locator.cpp]: See #242 for more details. As I mentioned it conflicts with it, still I don't want to copy/paste changes.

File: src/service/locator.cpp (hunk 1 above)
Comment (3Hren) [inline src/service/locator.cpp]: Whatever, don't see much difference.
@kobolog thoughts?

File: src/service/locator.cpp (hunk 2 above)
Comment (3Hren) [inline src/service/locator.cpp]: Again whatever.

Also this change partially conflicts with #246, I'll rebase if necessary.

File: src/service/locator.cpp [hunk 4]
```diff
@@ -280,19 +280,50 @@ locator_cfg_t::locator_cfg_t(const std::string& name_, const dynamic_t& root):
...
+        return {value.first, value.second.all()};
+    });
+
+    return results;
+}
+
+streamed<results::routing>
+locator_t::routing_t::notify(const std::string& ruid, const results::routing& results, policy_t policy) {
+    auto stream = routers.apply([&](router_map_t& mapping) -> streamed<results::routing> {
+        if(mapping.count(ruid) == 0 || ((policy == policy_t::replace) && mapping.erase(ruid))) {
```
Comment (antmat) [inline src/service/locator.cpp]: in case of replace we do 2 searches in map. It's not critical, but we can use find here and erase by iterator.

**Comment by antmat** [comment]
LGTM in general

File: src/service/locator.cpp (hunk 4 above)
Comment (3Hren) [inline src/service/locator.cpp]: Actually we do 3 searches (see following insert). This isn't critical here.

**Comment by antmat** [comment]
During routing group update we do we following:
  * iterate over routers and store their ruids in vector
  * iterate over vector of ruids, find corresponding upstream in routers and write update to upstream.

Looks like it was done to reuse existing on_routing code, however it looks like it performs many unnescessary operations.

File: include/cocaine/detail/service/locator.hpp [hunk 5]
```diff
@@ -120,8 +117,32 @@ class locator_t:
...
+
+        routing_t(const logging::log_t& log):
+            log(log)
+        {}
+
+        results::routing
+        dump() const;
+
+        streamed<results::routing>
+        notify(const std::string& ruid, const results::routing& results, policy_t policy = policy_t::reuse);
```
Comment (antmat) [inline include/cocaine/detail/service/locator.hpp]: I would also drop default value here.

**Comment by 3Hren** [comment]
Partially you're right.

But here we perform write operation without locking.

Still there can be possible race condition when we failed to write message, then new router stream with the same id is registered, then we erase it as a result of previous failed write.

I don't know how it's important, let's ask the author (@kobolog, ping).

File: include/cocaine/detail/service/locator.hpp [hunk 6]
```diff
@@ -120,8 +117,32 @@ class locator_t:
     // Snapshots of the local service states. Synchronized with outgoing remote streams.
     std::map<std::string, results::resolve> m_snapshots;
```
Comment (kobolog) [inline include/cocaine/detail/service/locator.hpp]: This one is technically part of the routing_t, I believe.

**Comment by kobolog** [comment]
I don't think it's important, in this case performance > consistency: in case this RC happens, router will simply reconnect.

**Comment by kobolog** [comment]
Maybe it's a good time to extract all routing logic into locator/routing.hpp, and put it together with Ketama stuff?

**Comment by kobolog** [comment]
If you refactor & rebase it on current master, I'll merge it.

**Pull request closed without merging**